### PR TITLE
[Merged by Bors] - feat: port Topology.ContinuousFunction.Algebra

### PR DESCRIPTION
**Pull request by mans0954**
---

- [x] depends on: #3770

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mans0954** [comment]
I'm a bit stuck with `def continuousSubring` https://github.com/leanprover-community/mathlib4/pull/3332/files#diff-fea023be112f176cf1013bffd82df64e4459f2f386651e5f827e76184e48e0c1R492-R495

In Mathlib3 we have:
```
/-- The subring of continuous maps `α → β`. -/
def continuous_subring (α : Type*) (R : Type*) [topological_space α] [topological_space R]
  [ring R] [topological_ring R] : subring (α → R) :=
{ ..continuous_subsemiring α R,
  ..continuous_add_subgroup α R }
```
In Mathlib4, it is still the case that a topological ring is a topological semiring and a topological additive group:
```
example (R : Type _) [Ring R] [TopologicalSpace R] [TopologicalRing R] :
  TopologicalSemiring R := TopologicalSemiring.mk

example (R : Type _) [Ring R] [TopologicalSpace R] [TopologicalRing R] :
  TopologicalAddGroup R := TopologicalRing.to_topologicalAddGroup
```
Lean seems to be able to infer that a topological ring is a topological additive group automatically, but not that it is a topological semiring:
```
variable (γ : Type _) [TopologicalSpace γ]
variable (S : Type _) [TopologicalSpace S] [Ring S] [TopologicalRing S]

#check continuousAddSubgroup γ S
#check continuousSubsemiring γ S #failed to synthesize instance TopologicalSemiring S
```
Ultimately, I can't get the proof of
```
def continuousSubring (α : Type _) (R : Type _) [TopologicalSpace α] [TopologicalSpace R] [Ring R]
    [TopologicalRing R] : Subring (α → R) :=
```
right in Mathlib4.

**Comment by mans0954** [comment]
> I'm a bit stuck with `def continuousSubring`

Think I've fixed this by weakening the hypothesis of `continuous_subsemiring`.

Unfortunately this doesn't solve the problem for `instance ... NonAssocRing C(α, β)`.

**Comment by j-loreaux** [comment]
Note: I have just now incorporate a change which weakens the type class assumptions for `ContinuousMap.module'` to allow `Semiring`. This mimics the change in the not-yet-merged leanprover-community/mathlib#18868. Please update the SHA accordingly after that PR is merged so that we don't incur forward-porting obligations / review.

**Comment by eric-wieser** [comment]
> Please update the SHA accordingly after that PR is merged so that we don't incur forward-porting obligations / review.

Done. There's now a new "latest mathport output" base commit to diff against.

File: Mathlib/Topology/ContinuousFunction/Algebra.lean [hunk 1]
```diff
@@ -134,28 +138,29 @@ theorem int_cast_apply [IntCast β] (n : ℤ) (x : α) : (n : C(α, β)) x = n :
...
+--#align continuous_map.has_nsmul ContinuousMap.hasNSMul
 
-@[norm_cast, to_additive]
+--@[norm_cast, to_additive]
+@[to_additive]
 theorem coe_pow [Monoid β] [ContinuousMul β] (f : C(α, β)) (n : ℕ) : ⇑(f ^ n) = f ^ n :=
   rfl
 #align continuous_map.coe_pow ContinuousMap.coe_pow
-#align continuous_map.coe_nsmul ContinuousMap.coe_nsmul
+--#align continuous_map.coe_nsmul ContinuousMap.coe_nsmul
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Algebra.lean:156]: These should not be commented out.

File: Mathlib/Topology/ContinuousFunction/Algebra.lean [hunk 2]
```diff
@@ -350,24 +356,24 @@ variable (α)
 
 /-- Composition on the left by a (continuous) homomorphism of topological monoids, as a
 `monoid_hom`. Similar to `monoid_hom.comp_left`. -/
-@[to_additive
-      "Composition on the left by a (continuous) homomorphism of topological `add_monoid`s,\nas an `add_monoid_hom`. Similar to `add_monoid_hom.comp_left`.",
-  simps]
-protected def MonoidHom.compLeftContinuous {γ : Type _} [Monoid β] [ContinuousMul β]
+--@[to_additive (attr := simps)
+--    "Composition on the left by a (continuous) homomorphism of topological `add_monoid`s,\nas an
+-- `add_monoid_hom`. Similar to `add_monoid_hom.comp_left`."]
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Algebra.lean:361]: Please don't just remove this docstring

File: Mathlib/Topology/ContinuousFunction/Algebra.lean [hunk 3]
```diff
@@ -220,20 +225,21 @@ theorem div_comp [Div γ] [ContinuousDiv γ] (f g : C(β, γ)) (h : C(α, β)) :
 
 -- ### "zpow" and "zsmul"
 instance hasZsmul [AddGroup β] [TopologicalAddGroup β] : SMul ℤ C(α, β)
-    where smul z f := ⟨z • f, f.Continuous.zsmul z⟩
+    where smul z f := ⟨SMul.smul z  f, f.continuous.zsmul z⟩
```
Comment (eric-wieser) [inline Mathlib/Topology/ContinuousFunction/Algebra.lean:226]: You should never write `SMul.smul`, as that's not what `•` is syntax for in Lean 4.

Comment (mans0954) [reply Mathlib/Topology/ContinuousFunction/Algebra.lean:227]: @eric-wieser thanks - still struggling with [this one](https://github.com/leanprover-community/mathlib4/pull/3332/files#diff-fea023be112f176cf1013bffd82df64e4459f2f386651e5f827e76184e48e0c1R843) though:

```
    use v x • (1 : C(α, 𝕜)) #
application type mismatch
  Exists.intro (v x • 1)
argument
  v x • 1
has type
  C(α, 𝕜) : outParam (Type (max u_1 u_2))
but is expected to have type
  ↑↑s : Type (max u_1 u_2)Lean 4
```

Comment (eric-wieser) [reply Mathlib/Topology/ContinuousFunction/Algebra.lean:227]: I have a mathlib3 PR in the works that cleans this proof up, which might be the easy way out here

Comment (eric-wieser) [reply Mathlib/Topology/ContinuousFunction/Algebra.lean:227]: #18888

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3770
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3770~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**